### PR TITLE
fix: ensure atomicity in skill installation and add defensive dependency loading

### PR DESCRIPTION
**Pull request by ChandanKT-git**
## Pull Request description

This PR addresses a critical race condition and lack of runtime validation in the `SkillRegistry` regarding skill dependencies.

**Changes:**
1. **Atomicity Fix**: In `install_skill` and `update_skill`, the logic was swapped to run `pip install` **before** updating the `state.json` file. This ensures that a skill is only marked as "Installed" if its environment setup actually succeeded.
2. **Defensive Loading**: Implemented a `_verify_dependencies` check in the `load()` path. It uses `importlib.util.find_spec` to verify that all packages listed in the skill's manifest are actually present in the current Python environment before attempting an import.
3. **User-Friendly Errors**: Introduced `MissingDependencyError`, which provides a clear message to the user/UI about which packages are missing, replacing raw `ModuleNotFoundError` crashes.

## How to test these changes

I have added a comprehensive new test suite: `tests/pipeline/test_registry_dependency_safety.py`.

To run the new tests:
```bash
python -m pytest tests/pipeline/test_registry_dependency_safety.py -v
```

To verify no regressions:
```bash
python -m pytest tests/test_registry.py tests/pipeline/test_registry_channels.py -v
```

## Pull Request checklists

This PR is a:

- [x] bug-fix
- [ ] new feature
- [ ] maintenance

About this PR:

- [x] it includes tests.
- [ ] the tests are executed on CI.
- [ ] the tests generate log file(s) (path).
- [x] pre-commit hooks were executed locally.
- [ ] this PR requires a project documentation update.

Author's checklist:

- [x] I have reviewed the changes and it contains no misspelling.
- [x] The code is well commented, especially in the parts that contain more complexity.
- [x] New and old tests passed locally.

## Additional information

The `_verify_dependencies` method includes logic to strip version specifiers (e.g., `package>=1.0` -> `package`) and normalize hyphens to underscores to ensure compatibility with `importlib` lookups.

### CODE REVIEW

**Comment by xmnlab** [comment]
hi @ChandanKT-git , thanks for working on that.

Thanks for this PR — the overall direction makes sense. Moving dependency installation before `_save_state()` is a good fix for the state consistency issue, and the new `MissingDependencyError` makes the load failure much easier to understand.

I would hold this before merge for one important issue in `_verify_dependencies()`.

The current implementation converts dependency names into import names by stripping specifiers and replacing `-` with `_`, then checking `importlib.util.find_spec()`. That will produce false missing-dependency errors for many valid Python packages where the distribution name and import name differ, for example `pillow -> PIL`, `python-magic -> magic`, `PyYAML -> yaml`, `python-dateutil -> dateutil`, `beautifulsoup4 -> bs4`, `opencv-python -> cv2`, and `scikit-learn -> sklearn`.

I think this should validate installed distributions instead of guessed import names, using `packaging.requirements.Requirement` plus `importlib.metadata.version()` / `PackageNotFoundError`. That would also let us respect version specifiers and environment markers more correctly.

It would also be great to add tests for at least one distribution/import-name mismatch, such as `pillow` or `python-dateutil`, so this does not regress later.

The rest of the PR looks good conceptually, but CI is currently red because the linter job fails on a mypy internal error. Even if that failure is probably not caused by this PR, I would not merge while the build workflow is failing.

maybe you can explorer someting like this:

```
from importlib.metadata import PackageNotFoundError, version

from packaging.requirements import InvalidRequirement, Requirement


@staticmethod
def _verify_dependencies(skill_id: str, dependencies: list[str]) -> None:
    missing: list[str] = []

    for dep in dependencies:
        try:
            req = Requirement(dep)
        except InvalidRequirement:
            missing.append(dep)
            continue

        if req.marker is not None and not req.marker.evaluate():
            continue

        try:
            installed_version = version(req.name)
        except PackageNotFoundError:
            missing.append(dep)
            continue

        if req.specifier and not req.specifier.contains(
            installed_version,
            prereleases=True,
        ):
            missing.append(dep)

    if missing:
        raise MissingDependencyError(skill_id, missing)

```

**Comment by xmnlab** [comment]
also please rebase your branch on top of the upstream main branch

**Comment by esloch** [comment]
Hi @ChandanKT-git. Thanks for working on this! Overall, I think this is a good improvement.

I just have one concern about `_verify_dependencies()`. Using `importlib.util.find_spec()` assumes the package name matches the import name, but that's not always true (e.g. `pillow` → `PIL`, `PyYAML` → `yaml`, `scikit-learn` → `sklearn`).

I think it would be more robust to validate installed distributions using `importlib.metadata` together with `packaging.Requirement`, which also handles version specifiers correctly.

It would also be nice to add a regression test covering one of these cases.
